### PR TITLE
Feature/git lib

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

File: src/docfx/docfx.csproj [hunk 1]
```diff
@@ -5,6 +5,12 @@
     <TargetFramework>netcoreapp2.0</TargetFramework>
     <LangVersion>Latest</LangVersion>
     <CodeAnalysisRuleSet>../../stylecop.ruleset</CodeAnalysisRuleSet>
+    <AssemblyName>Microsoft.Docs</AssemblyName>
+    <RootNamespace>Microsoft.Docs</RootNamespace>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Comment (yufeih) [inline src/docfx/docfx.csproj]: Remove this condition

File: src/docfx/docfx.csproj [hunk 2]
```diff
@@ -14,6 +20,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.0" PrivateAssets="All" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Octokit" Version="0.29.0" />
```
Comment (yufeih) [inline src/docfx/docfx.csproj]: Split GitHub and Git as 2 PRs

File: src/docfx/lib/git/Git.cs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+using FileMap = System.Collections.Generic.Dictionary<string, (string file, string parent)>;
+using TreeMap = System.Collections.Concurrent.ConcurrentDictionary<long, System.Collections.Generic.Dictionary<string, long>>;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide git operations
+    /// </summary>
+    public static class Git
```
Comment (yufeih) [inline src/docfx/lib/git/Git.cs]: `GitUtil` to be consistent with other utils?

File: src/docfx/lib/git/NativeMethods.cs [hunk 4]
```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Docs
+{
+    internal static class NativeMethods
```
Comment (yufeih) [inline src/docfx/lib/git/NativeMethods.cs]: can we disable stylecop entirely for this file and Git.cs?

We should keep the function names and struct fields the same as LibGit2, like `git_object_id`

File: stylecop.ruleset [hunk 5]
```diff
@@ -23,6 +23,12 @@
     <Rule Id="SA1503" Action="None" />
     <!-- SA1513ClosingBraceMustBeFollowedByBlankLine -->
     <Rule Id="SA1513" Action="None" />
+    <!-- CA2007DoNotDirectlyAwaitATaskAnalyzer -->
```
Comment (yufeih) [inline stylecop.ruleset]: put CA rules inside <Rules AnalyzerId="Microsoft.Analyzers.ManagedCodeAnalysis" RuleNamespace="Microsoft.Rules.Managed">

File: test/docfx.Test/docfx.Test.csproj [hunk 6]
```diff
@@ -5,6 +5,12 @@
     <IsPackable>false</IsPackable>
     <LangVersion>Latest</LangVersion>
     <CodeAnalysisRuleSet>../../stylecop.ruleset</CodeAnalysisRuleSet>
+    <AssemblyName>Microsoft.Docs.Test</AssemblyName>
+    <RootNamespace>Microsoft.Docs.Test</RootNamespace>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Comment (yufeih) [inline test/docfx.Test/docfx.Test.csproj]: Remove this condition and merge it with the above property group

File: src/docfx/lib/PathUtil.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    public static class PathUtil
+    {
+        /// <summary>
+        /// Determines if a file or path is inside a folder. Input should both be normalized.
+        /// </summary>
+        public static bool Inside(string folder, string path)
+        {
+            Debug.Assert(folder.Length > 0 && folder[folder.Length - 1] == '/');
+
+            return folder == "./" || path.StartsWith(folder, StringComparison.OrdinalIgnoreCase);
```
Comment (herohua) [inline src/docfx/lib/PathUtil.cs]: >OrdinalIgnoreCase [](start = 78, length = 17)

should use platform dependent comparison?

**Comment by herohua** [comment]
The coding style here seems not the same as v2, are we going to follow the same style as v2 or will we define a new set of styles for v3? @vicancy, @zhyan, any thoughts?

File: src/docfx/lib/git/Github.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                    }
+                }
+            }
+            return res;
+        }
+
+        private static async Task<(GitHubUser author, GitHubUser committer)> GetUserFromCommit(string owner, string name, string reference, GitHubClient github)
+        {
+            const int retry = 5;
+            for (var i = 0; i <= 5; i++)
```
Comment (herohua) [inline src/docfx/lib/git/Github.cs]: >5 [](start = 33, length = 1)

retry

**Pull request closed without merging**